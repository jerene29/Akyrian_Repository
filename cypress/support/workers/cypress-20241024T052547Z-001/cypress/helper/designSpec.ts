import { GetStudyRevisionListDocument } from '../../src/graphQL/generated/graphql';
import { mockUserDataAdmin } from '../../src/constant/testFixtures';

type SpecNumber = `S-${number}-${number}`;

type SpecInfo = {
  // e.g S-09-010
  spec: SpecNumber;
  // title e.g headerSection, combined description becomes S-09-010 headerSection, but optional
  title?: string;
};

type Options = {
  loginAs?: string;
  skipTest?: boolean;
};

function resolveDescription(description: SpecNumber | SpecInfo) {
  return typeof description === 'string' ? description : `${description.spec} ${description.title}`;
}

const aliasGetStudyRevisionListDocument =
  'name' in GetStudyRevisionListDocument.definitions[0]
    ? GetStudyRevisionListDocument.definitions[0].name?.value
    : 'GetStudyRevisionListDocument';
/*
example usage
designSpec('S-19-2', () => {

  describe('...', ...)

  it('checks something...', () => {...})
});
*/
export function designSpec(
  description: SpecNumber | SpecInfo,
  callback: () => void,
  options?: Options,
) {
  before(() => {
    cy.intercept('POST', '/graphql', (req) => {
      if (req.body.operationName === aliasGetStudyRevisionListDocument) {
        req.alias = req.body.operationName;
      }
    });
    cy.beforeSetup(mockUserDataAdmin);
    cy.clearLocalStorageSnapshot();
    cy.saveLocalStorage();
    cy.wait(`@${aliasGetStudyRevisionListDocument}`);
  });
  if (options?.skipTest) {
    describe.skip(resolveDescription(description), callback);
  } else if (typeof description !== 'string' && description.title) {
    describe(description.spec, () => {
      describe(String(description.title), callback);
    });
  } else {
    describe(resolveDescription(description), callback);
  }
}
