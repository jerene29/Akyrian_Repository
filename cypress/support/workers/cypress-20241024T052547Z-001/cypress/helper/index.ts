export * from './designSpec';
export * from './object';