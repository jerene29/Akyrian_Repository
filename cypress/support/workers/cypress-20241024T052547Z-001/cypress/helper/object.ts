/**
 * to shorten the tedious '[data-cy=selector]' selector.
 * just put the name of data-cy attributed element.
 * -- d stands for document
 *
```
d`addButton`
```
 */
export function d(args: TemplateStringsArray) {
  return `[data-cy=${args.join('')}]`;
}
