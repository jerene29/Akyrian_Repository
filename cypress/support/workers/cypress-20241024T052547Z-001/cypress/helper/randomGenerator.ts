import moment = require("moment");

export const randomAlphaNumeric = (length: number) => {
  const chars = '0123456789abcdefghijklmnopqrstuvwxyzABCDEFGHIJKLMNOPQRSTUVWXYZ'
  let result = '';
  for (let i = length; i > 0; --i) result += chars[Math.floor(Math.random() * chars.length)];
  return result;
}

export const randomAlphabet = (length: number) => {
  const chars = 'ABCDEFGHIJKLMNOPQRSTUVWXYZ'
  let result = '';
  for (let i = length; i > 0; --i) result += chars[Math.floor(Math.random() * chars.length)];
  return result;
}

export const randomDate = () => {
  const start = new Date(2012, 0, 1)
  const end = new Date()
  const date = new Date(start.getTime() + Math.random() * (end.getTime() - start.getTime()))
  return moment(date);
}