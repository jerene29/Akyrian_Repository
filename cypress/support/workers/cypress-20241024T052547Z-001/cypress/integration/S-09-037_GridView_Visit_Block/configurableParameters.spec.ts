import { d, designSpec } from '../../helper';

const options = {
  loginAs: 'admin@example.com',
};

designSpec(
  {
    spec: 'S-09-037',
    title: 'Configurable Parameters on Visit Edit mode',
  },
  () => {
    before(() => {
      // rely on seed data
      cy.visit('/study/study1revisionDev2e/templateScreeningRev2eDev/consentRev2eDevId1');
      cy.shouldBeVisible(d`visit-name`);
      cy.get(d`grid-view`).click();
    });

    const EDIT_TARGET = d`editableIntervisitInterval-0-3`;
    const INPUT_TARGET = d`relativeDayOffsetInput-0-3`;

    context('Intervisit Spatial Map', () => {
      it('should be visible', () => {
        cy.get(d`grid-view`).click();
        cy.hover(d`headerCell-0-3`);
        cy.shouldBeVisible(d`intervisitInterval-0-2`);
        cy.shouldBeVisible(EDIT_TARGET);
      });

      it('cannot be left blank', () => {
        cy.hover(d`headerCell-0-3`);
        cy.shouldBeVisible(EDIT_TARGET).click();
        cy.get(INPUT_TARGET).click().clear();
        cy.getTextSnapshot(d`cellErrorPopover`);
      });

      it('must be an integer', () => {
        cy.hover(d`headerCell-0-3`);
        cy.shouldBeVisible(EDIT_TARGET).click();
        cy.get(INPUT_TARGET).click().clear().type('YOLO');
        cy.getTextSnapshot(d`cellErrorPopover`);
        // Remove this when 5000 validation is re-activated
        cy.get(INPUT_TARGET).click().clear().type('{esc}{esc}');
      });

      // TODO: Revert this once max value is decided.
      // Changing this from 500 to 5000, since it was the max value. But ignoring this for now based on task below
      // Reference: https://ikeguchiholdings.monday.com/boards/1959201049/pulses/3406468584
      xit('min is 1 max is 5000', () => {
        cy.hover(d`headerCell-0-3`);
        cy.shouldBeVisible(EDIT_TARGET).click();
        cy.get(INPUT_TARGET).click().clear().type('9999');
        cy.getTextSnapshot(d`cellErrorPopover`);
        cy.get(INPUT_TARGET).click().clear().type('{esc}{esc}');
      });
    });

    context('visit window', () => {
      it('should be visible', () => {
        cy.hover(d`headerCell-0-3`);
        cy.shouldBeVisible(d`visitWindowBefore-0-3`);
        cy.shouldBeVisible(d`visitWindowAfter-0-3`);
      });

      it('cannot be left blank', () => {
        cy.hover(d`headerCell-0-3`);
        cy.shouldBeVisible(d`visitWindowBefore-0-3`).click();
        // impossible since value is default to 0.
        cy.shouldBeVisible(d`editableVisitWindowBefore-0-3`)
          .focus()
          .clear()
          // also impossible since input filters non numerical characters
          .type('hello');
      });

      it('max to 100', () => {
        cy.shouldBeVisible(d`editableVisitWindowBefore-0-3`)
          .focus()
          .clear()
          .type('999');
        cy.getTextSnapshot(d`cellErrorPopover`);
      });

      it('checks if the system always places negative (-)', () => {
        cy.shouldBeVisible(d`visitWindowBefore-0-4`).contains('-');
      });
    });

    context('visit name', () => {
      it('should be visible', () => {
        cy.get('body').type('{esc}');
        cy.hover(d`headerCell-0-3`);
        cy.shouldBeVisible(d`visitNameBox`);
      });

      it('cannot be left blank', () => {
        cy.hover(d`headerCell-0-3`);
        cy.shouldBeVisible(d`visitNameBox`).click();

        cy.shouldBeVisible(d`visitNameInput`)
          .focus()
          .clear();
        cy.getTextSnapshot(d`cellErrorPopover`);
      });

      it('validate special characters', () => {
        cy.hover(d`headerCell-0-3`);
        cy.shouldBeVisible(d`visitNameBox`).click();

        cy.shouldBeVisible(d`visitNameInput`)
          .focus()
          .clear()
          .type('...(*');
        cy.getTextSnapshot(d`cellErrorPopover`);
      });

      it('max is 25', () => {
        // TODO: make utils cos they're basically the same sequence of action.
        cy.hover(d`headerCell-0-3`);
        cy.shouldBeVisible(d`visitNameBox`).click();

        cy.shouldBeVisible(d`visitNameInput`)
          .focus()
          .clear()
          .type(
            'akwkwkwkakwakwakwkawkakwakwakwkawkakwakwkakwaowakokaokwoakwoakwoakowkaowkaoakwowakoakwowakoawkowkwkwaokwoakwoakowkoakoawkwoakaokwoakwo',
          );
        cy.getTextSnapshot(d`cellErrorPopover`);
      });
    });

    context('visit target (dayOffset)', () => {
      const typeDayOffset = (input?: string) => {
        cy.get('body').type('{esc}');
        cy.hover(d`headerCell-0-3`);
        cy.shouldBeVisible(d`dayOffsetBox`).click();
        if (!input) {
          cy.shouldBeVisible(d`dayOffsetInput`)
            .focus()
            .clear();
        } else {
          cy.shouldBeVisible(d`dayOffsetInput`)
            .focus()
            .clear()
            .type(input);
        }
      };
      it('cannot be left blank', () => {
        typeDayOffset();
        cy.getTextSnapshot(d`cellErrorPopover`);
      });
      // TODO: Revert this once max value is decided, ignoring this for now based on task below
      // Reference: https://ikeguchiholdings.monday.com/boards/1959201049/pulses/3406468584
      xit('max to 5000', () => {
        typeDayOffset('9999');
        cy.getTextSnapshot(d`cellErrorPopover`);
      });

      it('cannot have duplicate dayOffset', () => {
        typeDayOffset('7');
        cy.getTextSnapshot(d`cellErrorPopover`);
      });

      it('editing visit target will automatically update the visit interval', () => {
        // before
        cy.get('body').type('{esc}');
        cy.hover(d`headerCell-0-3`);
        typeDayOffset('10');
        // after
        cy.getTextSnapshot(d`editableIntervisitInterval-0-3`);
      });

      it('editing interval value will automatically update the visit target', () => {
        // before
        cy.get('body').type('{esc}');
        cy.hover(d`headerCell-0-3`);
        cy.get(d`editableIntervisitInterval-0-3`).click();
        cy.get(d`relativeDayOffsetInput-0-3`)
          .focus()
          .clear()
          .type('3');
        // after
        cy.getTextSnapshot(d`dayOffsetInactiveInput`);
      });
    });
  },
  options,
);
