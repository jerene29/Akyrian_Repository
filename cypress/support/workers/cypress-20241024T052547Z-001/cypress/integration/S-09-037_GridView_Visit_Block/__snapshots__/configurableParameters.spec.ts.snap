exports[`S-09-037 > Configurable Parameters on Visit Edit mode > Intervisit Spatial Map > cannot be left blank #0`] =
{
  "snapshot": "this field cannot be left blank"
};

exports[`S-09-037 > Configurable Parameters on Visit Edit mode > Intervisit Spatial Map > cannot be left blank #1`] =
{
  "snapshot": ""
};

exports[`S-09-037 > Configurable Parameters on Visit Edit mode > Intervisit Spatial Map > must be an integer #0`] =
{
  "snapshot": "this field cannot be left blank"
};

exports[`S-09-037 > Configurable Parameters on Visit Edit mode > Intervisit Spatial Map > must be an integer #1`] =
{
  "snapshot": ""
};

exports[`S-09-037 > Configurable Parameters on Visit Edit mode > Intervisit Spatial Map > min is 1 max is 500 #0`] =
{
  "snapshot": "max value is 500"
};

exports[`S-09-037 > Configurable Parameters on Visit Edit mode > Intervisit Spatial Map > min is 1 max is 500 #1`] =
{
  "snapshot": ""
};

exports[`S-09-037 > Configurable Parameters on Visit Edit mode > visit window > max to 100 #0`] =
{
  "snapshot": "max value is 100"
};

exports[`S-09-037 > Configurable Parameters on Visit Edit mode > visit window > max to 100 #1`] =
{
  "snapshot": ""
};

exports[`S-09-037 > Configurable Parameters on Visit Edit mode > visit name > cannot be left blank #0`] =
{
  "snapshot": "this field cannot be left blank"
};

exports[`S-09-037 > Configurable Parameters on Visit Edit mode > visit name > cannot be left blank #1`] =
{
  "snapshot": ""
};

exports[`S-09-037 > Configurable Parameters on Visit Edit mode > visit name > validate special characters #0`] =
{
  "snapshot": "can only contain letters, numbers and special characters: [space], -, _, #"
};

exports[`S-09-037 > Configurable Parameters on Visit Edit mode > visit name > validate special characters #1`] =
{
  "snapshot": ""
};

exports[`S-09-037 > Configurable Parameters on Visit Edit mode > visit name > max is 25 #0`] =
{
  "snapshot": "min 1 character, max 25 characters"
};

exports[`S-09-037 > Configurable Parameters on Visit Edit mode > visit name > max is 25 #1`] =
{
  "snapshot": ""
};

exports[`S-09-037 > Configurable Parameters on Visit Edit mode > visit target (dayOffset) > cannot be left blank #0`] =
{
  "snapshot": "this field cannot be left blank"
};

exports[`S-09-037 > Configurable Parameters on Visit Edit mode > visit target (dayOffset) > cannot be left blank #1`] =
{
  "snapshot": ""
};

exports[`S-09-037 > Configurable Parameters on Visit Edit mode > visit target (dayOffset) > max to 5000 #0`] =
{
  "snapshot": "max value is 5000"
};

exports[`S-09-037 > Configurable Parameters on Visit Edit mode > visit target (dayOffset) > max to 5000 #1`] =
{
  "snapshot": ""
};

exports[`S-09-037 > Configurable Parameters on Visit Edit mode > visit target (dayOffset) > cannot have duplicate dayOffset #0`] =
{
  "snapshot": "cannot have more than one visit on the same day"
};

exports[`S-09-037 > Configurable Parameters on Visit Edit mode > visit target (dayOffset) > cannot have duplicate dayOffset #1`] =
{
  "snapshot": ""
};

exports[`S-09-037 > Configurable Parameters on Visit Edit mode > visit target (dayOffset) > editing visit target will automatically update the visit interval #0`] =
{
  "snapshot": "3 days"
};

exports[`S-09-037 > Configurable Parameters on Visit Edit mode > visit target (dayOffset) > editing interval value will automatically update the visit target #0`] =
{
  "snapshot": "10"
};
